dev/test/bulk-writer.ts

```diff
@@ -0,0 +1,502 @@
+  let bulkWriter: BulkWriter;
+  let requestCounter: number;
+  let opCount: number;
+  const flushDeferred = new Deferred<void>();
+  let flushCounter = 0;
+
+  beforeEach(() => {
+    requestCounter = 0;
+    opCount = 0;
+  });
+
+  function incrementOpCount(): void {
+    opCount++;
+  }
+
+  function verifyOpCount(expected: number): void {
+    expect(opCount).to.equal(expected);
+  }
+
+  function setOp(doc: string, value: string): BulkWriterRequest {
+    return set({
+      document: document(doc, 'foo', value),
+    }) as BulkWriterRequest;
+  }
+
+  function createOp(doc: string, value: string): BulkWriterRequest {
+    return create({
+      document: document(doc, 'foo', value),
+    }) as BulkWriterRequest;
+  }
+
+  function updateOp(doc: string, value: string): BulkWriterRequest {
+    const updateRequest = update({
+      document: document(doc, 'foo', value),
+      mask: updateMask('foo'),
+    }) as BulkWriterRequest;
+    delete updateRequest.writes[0].currentDocument;
+    return updateRequest;
+  }
+
+  function deleteOp(doc: string): BulkWriterRequest {
+    return remove(doc) as BulkWriterRequest;
+  }
+
+  function joinRequests(requests: BulkWriterRequest[]): BulkWriterRequest {
+    return {
+      writes: requests.map(req => req.writes[0]),
+    };
+  }
+
+  function successResponse(seconds: number): {[key: string]: {}} {
+    return {
+      updateTime: {
+        nanos: 0,
+        seconds,
+      },
+      code: Status.OK,
+    };
+  }
+
+  function failResponse(): {[key: string]: {} | null} {
+    return {
+      updateTime: null,
+      code: Status.UNAVAILABLE,
+    };
+  }
+
+  function createResponse(
+    responses: Array<{[key: string]: {} | null}>
+  ): BulkWriterResponse {
+    const writeResults = responses.map(response => ({
+      updateTime: response.updateTime,
+    }));
+    const status = responses.map(response => ({code: response.code!}));
+    return {
+      writeResults,
+      status,
+    };
+  }
+
+  function instantiateInstance(
+    mock: RequestResponse[],
+    manualFlush = false
+  ): Promise<void> {
+    const overrides: ApiOverride = {
+      batchWrite: async request => {
+        expect(request).to.deep.eq({
+          database: `projects/${PROJECT_ID}/databases/(default)`,
+          writes: mock[requestCounter].request.writes,
+        });
+        if (manualFlush) {
+          flushCounter++;
+
+          // This expect statement is used to test that only one request is
+          // made at a time.
+          expect(flushCounter).to.equal(1);
+          await flushDeferred.promise;
+          flushCounter--;
+        }
+        const responsePromise = response({
+          writeResults: mock[requestCounter].response.writeResults,
+          status: mock[requestCounter].response.status,
+        });
+        requestCounter++;
+        return responsePromise;
+      },
+    };
+    return createInstance(overrides).then(firestoreClient => {
+      firestore = firestoreClient;
```

Comment:
Have you considered returning `firestore` instead of setting global state? The consuming code would look more natural:

```
  const firestore = await instantiateInstance([
      {
        request: setOp('doc', 'bar'),
        response: createResponse([successResponse(2)]),
      },
    ]);
}

```

or:

```
  const {firestore, bulkWriter}  = await instantiateInstance([
      {
        request: setOp('doc', 'bar'),
        response: createResponse([successResponse(2)]),
      },
    ]);
```

Reply:
Sounds good.